### ISSUE TITLE
Custom CSS/LESS box not taking precedence

### ISSUE THREAD

**Issue by canstudios-nicolaw** [adaptlearning/adapt_authoring#2368]
## Expected Behaviour
The rules entered into the custom CSS/LESS box on a course should take precedence over all other styling rules in the theme.

## Actual Behaviour
If I have Vanilla 3.1.0 and I am using Framework 4 then the rules I enter into the CSS box don't seem to take the same precedence that they used to in v2. This is leading to increased use of !important.

## Steps to Reproduce
1. Create a course and in the custom less box add `.boxmenu-menu .menu-item-inner {
   background-color: blue;
}`
1. Preview the course
1. The menu items should be blue, but instead they are Vanilla's standard colour. In the browser's inspector I can see that Vanilla's rule has taken precedence over my own.

## Versions
  - Authoring Tool Version: 0.8.1
  - Framework Version: 4.2.0

**Comment by Link2Twenty** [comment]
Don't things in that box just follow the [css specificity rules](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity)? Personally, I'd have thought that is better than all user-entered CSS acting as though it had `!important` suffix.

To take your example adding another layer of specificity to make the rule more specific than the one in the theme fixes the issue.

```css 
.boxmenu-menu .menu-item .menu-item-inner { background-color: blue; }
```

**Comment by canstudios-nicolaw** [comment]
Yes they do, but the idea of that field was that it could be used to overwrite styles in the theme (and this is how it used to work in v2). This behaviour has changed in v4 and the change wasn't by design - I believe it has been caused largely by the restructure of Vanilla in v4.

The rules in this box should be applied after the rules in the theme. In v4, the rules in this box are applied before some rules in the theme and after others, this is confusing and inconsistent.

Yes writing more specific rules will get around the issue but didn't used to be required and can become quite cumbersome - particularly for someone who does not have much experience with Less as they used to be able to copy a theme rule and change the value, now they must work out how to make the rule more specific in some cases.

**Comment by Link2Twenty** [comment]
That makes sense, having the rules appear after the theme rules will still only affect styles in cases where the specificity matches (or is greater than) the original rule though. I guess, however, this was the case before so people are used to it 🙂

**Comment by canstudios-nicolaw** [comment]
I've done some digging into grunts less compilation task, and in v4 the Less files get pulled in in the following order

```
@import 'src/core/less/base.less';
@import 'src/core/less/reset.less';
@import 'src/core/less/scrolling.less';
@import 'src/components/adapt-contrib-hotgraphic/less/hotgraphic.less';
@import 'src/components/adapt-contrib-narrative/less/narrative.less';
@import 'src/components/adapt-contrib-text/less/text.less';
@import 'src/menu/adapt-contrib-boxMenu/less/boxmenu.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/adapt-defaults.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/article.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/block.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/buttons.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/component.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/drawer.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/fonts.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/icons.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/loading.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/navigation.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/notify.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/page.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/shadow.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/transitions.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/wrapper.less';
@import 'src/theme/adapt-contrib-vanilla/less/custom/theme-extras.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/colors.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/fonts.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/generic.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/paddings.less';
@import 'src/theme/adapt-contrib-vanilla/less/zzzzz/1-themeVariables.less';
@import 'src/theme/adapt-contrib-vanilla/less/zzzzz/2-customStyles.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-accordion/accordion.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-assessmentResults/assessmentResults.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-boxMenu/boxmenu.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-gmcq/gmcq.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-hotgraphic/hotgraphic.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-matching/matching.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-mcq/mcq.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-media/media.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-narrative/narrative.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-pageLevelProgress/pageLevelProgress.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-resources/resources.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-slider/slider.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-textinput/textinput.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-trickle/trickle.less';
```

Clearly there is an issue in that the Less added through the AT (things in the zzzzz directory) gets imported before everything inside the theme plugins directory.

As far as I am aware, we only ever put AT Less inside a zzzzz directory so that it would be imported last, and we only ever put it inside the theme so that it would just get pulled in and compiled by the existing grunt tasks.

I suggest we modify the grunt task on the framework side, and modify the way that AT Less gets saved on the AT side.

Proposed changes:
Currently the grunt task gets less files in the following order

```
src: [
                    '<%= sourcedir %>components/**/*.less',
                    '<%= sourcedir %>extensions/**/*.less',
                    '<%= sourcedir %>menu/<%= menu %>/**/*.less',
                    '<%= sourcedir %>theme/<%= theme %>/**/*.less'
                ],
```

I think the best option would be to move the AT Less outside of the theme directory, so that an exported course would have the following directories: components, core, course, extensions, menu, theme, customStyles. The customStyles directory is new and would contain the same two Less files that can currently be found in zzzzz.

Then the grunt task could be modified to compile files in the following order:

```
src: [
                    '<%= sourcedir %>components/**/*.less',
                    '<%= sourcedir %>extensions/**/*.less',
                    '<%= sourcedir %>menu/<%= menu %>/**/*.less',
                    '<%= sourcedir %>theme/<%= theme %>/**/*.less',
                    '<%= sourcedir %>customStyles/**/*.less'
                ],
```

We would also need to make changes to the AT so that custom styles got saved to this new location rather than into the zzzzz directory.

Pros:
- Custom styles will always be applied after theme styles, regardless of what is in the theme
- One line change to framework side
- We get rid of the name zzzzz

Cons:
- We have to change a framework grunt task to support the AT

Does this sound like a reasonable approach?

**Comment by oliverfoster** [comment]
Can we just change the sort functions so that they sort correctly?

https://github.com/adaptlearning/adapt_framework/blob/master/grunt/config/less.js#L5-L56

**Comment by canstudios-nicolaw** [comment]
Ok I will take a look into this

**Comment by canstudios-nicolaw** [comment]
Morning @oliverfoster - I have temporarily removed the offending line you mention and now the order seems to be correct (shown below). I'm struggling to understand the second part of your sentence -  "except where the smaller path is the directory root of the larger path", please could you give me an example of this?

I tried adding a new file `@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-accordion/test/accordiontest.less';` to the theme to see if this was sorted incorrectly but I believe it is in the right place. (It is in the below output).

Thank you


```
@adapt-device-small:520;
@adapt-device-medium:760;
@adapt-device-large:900;
@import 'src/core/less/base.less';
@import 'src/core/less/reset.less';
@import 'src/core/less/scrolling.less';
@import 'src/components/adapt-contrib-hotgraphic/less/hotgraphic.less';
@import 'src/components/adapt-contrib-narrative/less/narrative.less';
@import 'src/components/adapt-contrib-text/less/text.less';
@import 'src/menu/adapt-contrib-boxMenu/less/boxmenu.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/adapt-defaults.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/article.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/block.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/buttons.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/component.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/drawer.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/fonts.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/icons.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/loading.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/navigation.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/notify.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/page.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/shadow.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/transitions.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/wrapper.less';
@import 'src/theme/adapt-contrib-vanilla/less/custom/theme-extras.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-accordion/accordion.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-accordion/test/accordiontest.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-assessmentResults/assessmentResults.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-boxMenu/boxmenu.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-gmcq/gmcq.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-hotgraphic/hotgraphic.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-matching/matching.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-mcq/mcq.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-media/media.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-narrative/narrative.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-pageLevelProgress/pageLevelProgress.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-resources/resources.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-slider/slider.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-textinput/textinput.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-trickle/trickle.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/colors.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/fonts.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/generic.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/paddings.less';
@import 'src/theme/adapt-contrib-vanilla/less/zzzzz/1-themeVariables.less';
@import 'src/theme/adapt-contrib-vanilla/less/zzzzz/2-customStyles.less';
```

**Comment by oliverfoster** [comment]
I just meant where:

a: src/theme/adapt-contrib-vanilla/less/plugins/
b: src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-mcq/mcq.less'

a clearly precedes

Whereas:

a: src/theme/adapt-contrib-vanilla/less/zzzzz/
b: src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-mcq/mcq.less

b clearly precedes despite being longer.

------------------------------------------------------


In your above case, can you check that a new file `src/theme/adapt-contrib-vanilla/less/variables.less` precedes `src/theme/adapt-contrib-vanilla/less/variables/colors.less`?

We'd like all immediate files to precede folders, and both files and folders should be alphabetically sorted.

**Comment by canstudios-nicolaw** [comment]
Yes variables.less precedes variables/colors.less in this scenario. I will make an equivalent issue on the framework side and raise a PR shortly.

```
@adapt-device-small:520;
@adapt-device-medium:760;
@adapt-device-large:900;
@import 'src/core/less/base.less';
@import 'src/core/less/reset.less';
@import 'src/core/less/scrolling.less';
@import 'src/components/adapt-contrib-hotgraphic/less/hotgraphic.less';
@import 'src/components/adapt-contrib-narrative/less/narrative.less';
@import 'src/components/adapt-contrib-text/less/text.less';
@import 'src/menu/adapt-contrib-boxMenu/less/boxmenu.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/adapt-defaults.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/article.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/block.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/buttons.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/component.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/drawer.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/fonts.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/icons.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/loading.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/navigation.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/notify.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/page.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/shadow.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/transitions.less';
@import 'src/theme/adapt-contrib-vanilla/less/core/wrapper.less';
@import 'src/theme/adapt-contrib-vanilla/less/custom/theme-extras.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-accordion/accordion.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-assessmentResults/assessmentResults.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-boxMenu/boxmenu.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-gmcq/gmcq.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-hotgraphic/hotgraphic.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-matching/matching.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-mcq/mcq.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-media/media.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-narrative/narrative.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-pageLevelProgress/pageLevelProgress.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-resources/resources.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-slider/slider.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-textinput/textinput.less';
@import 'src/theme/adapt-contrib-vanilla/less/plugins/adapt-contrib-trickle/trickle.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/colors.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/fonts.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/generic.less';
@import 'src/theme/adapt-contrib-vanilla/less/variables/paddings.less';
@import 'src/theme/adapt-contrib-vanilla/less/zzzzz/1-themeVariables.less';
@import 'src/theme/adapt-contrib-vanilla/less/zzzzz/2-customStyles.less';
```

**Comment by oliverfoster** [comment]
thank you :)

**Comment by canstudios-nicolaw** [comment]
I've moved this onto the release milestone for now - I'm not sure if this should just be closed as it was an entirely framework side change

**Comment by canstudios-nicolaw** [comment]
Closing as this change was made framework side (released in 4.3.0)
